python/cugraph/structure/number_map.py

```diff
@@ -0,0 +1,263 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import cugraph
+
+class NumberMap:
+    class SingleGPU:
+        def __init__(self, df, col_names):
+            # TODO: Fill in implementation
+            pass
+
+        def append(self, df, col_names):
+            # TODO: Fill in implementation
+            pass
+
+        def compute(self):
+            # TODO: Fill in implementation
+            pass
+
+        def to_vertex_id(df, col_names):
+            # TODO: Fill in implementation
+            pass
+
+        def from_vertex_id(series):
+            # TODO: Fill in implementation
+            pass
+
+    class MultiGPU:
+        def __init__(self, ddf, col_names):
+            # TODO: Fill in implementation
+            pass
+
+        def append(self, ddf, col_names):
+            # TODO: Fill in implementation
+            pass
+
+        def compute(self):
+            # TODO: Fill in implementation
+            pass
+
+        def to_vertex_id():
+            # TODO: Fill in implementation
+            pass
+
+        def from_vertex_id():
+            # TODO: Fill in implementation
+            pass
+
+            
+    def __init__(self):
+        self.implementation = None
+
+    def compute_vals(column_names):
+        """
+        Helper function to compute internal column names based on external column names
+        """
+        return [str(i) for i in range(len(column_names))]
+
+    def from_dataframe(self, df, col_names):
+        """
+        Populate the numbering map with vertices from the specified
+        columns of the provided data frame.
+
+        Parameters
+        ----------
+        df : cudf.DataFrame or dask_cudf.DataFrame
+            Contains a list of external vertex identifiers that will be
+            numbered by the NumberMap class.
+        col_names: list of strings
+            This list of 1 or more strings contain the names
+            of the columns that uniquely identify an external
+            vertex identifier
+        """
+        if self.implementation is not None:
+            raise Exception('NumberMap is already populated')
+
+        if type(df) is cudf.DataFrame:
+            self.implementation = NumberMap.SingleGPU(df, col_names)
+        elif type(df) is dask_cudf.DataFrame:
```

Comment:
How does this integrate into _Consolidation_ workflows #809 ?

Reply:
Hadn't seen that particular thought process.  I would see consolidation as a separate step that could be done either before or after renumbering.  If consolidation reduces to single GPU then use the Single GPU renumbering.  If consolidation reduces to a cluster of GPUs then the output of consolidation would still be a dask_cudf data frame and it wouldn't really matter whether it had been consolidated or not.

Reply:
That's fair. Trying to see how would this look from the user's perspective when the dask dataframe and multiGPU and pipeline transfers to a single GPU.

To take the most out of consolidation, we would probably want to renumber first as renumbered IDs are expected to be smaller (and we will leverage multi GPU for renumbering). In this case, the consolidation feature should : 
- accept an un-renumbered distributed dask data frame
- use the multi-GPU renumber path from this PR
- consolidate and return a single GPU Di/Graph.

Just need to make sure that renumbering maps and related objects are compatible with this workflow.

Reply:
The NumberMap object would support this flow.

Reply:
Should we have a test?